mudslide/models.py

```diff
@@ -553,12 +562,205 @@ def dV(self, X: ArrayLike) -> ArrayLike:
+    
+        with open(outname, "r") as f:
+            data_dict = turboparse.parse_turbo(f)
+
+        parsed_nac_coupling = data_dict["egrad"]["coupling"]
+        
+        self.derivative_coupling = np.zeros((self.nstates(), self.nstates(), self.ndim()))
+        for i in range(self.nstates()):
+            for j in range(self.nstates()):
+                if i != j:
+                    for dct in parsed_nac_coupling:
+                        if (dct["bra_state"] == i) and (dct["ket_state"] == j):
+                            self.derivative_coupling[i][j] = np.array(dct["d/dR"]).reshape(self.ndim(), order="F")
+                            self.derivative_coupling[j][i] = self.derivative_coupling[i][j]
+
+        parsed_gradients = data_dict["rdgrad"]["gradient"]
+        parsed_gradients.extend(data_dict["egrad"]["gradient"])
+
+
+        self.gradients = np.zeros((self.nstates(),self.ndim()))
+        for state in self.states:
+            grads = []
+            for i in range(self.ndim() // 3):
+                grads.extend(
+                [
+                parsed_gradients[state]["d_dx"][i],
+                parsed_gradients[state]["d_dy"][i],
+                parsed_gradients[state]["d_dz"][i],
+                ]
+                        )
+            grads = np.array(grads)
+            self.gradients[state] = grads
+
+
+        self.force = -(self.gradients) 
+
+        parsed_energies = data_dict["egrad"]["excited_state"][0]["energy"]
+        energy = data_dict["ridft"]["energy"]
+        excited_energies = [
+            data_dict["egrad"]["excited_state"][i]["energy"] + energy for i in range(len(data_dict["egrad"]["excited_state"]))
+        ]
+
+        self.energies = [energy]
+        self.energies.extend(excited_energies)
+
+
+    def V(self, X: ArrayLike) -> ArrayLike:
+        out = np.zeros([self.nstates(), self.nstates()])
+        for i, e in enumerate(self.states):
+            out[i][i] = self.energies[e]
+        return out
+
+    def compute(self, X, couplings, gradients, reference):
+        """
+        Calls Turbomole/Turboparse to generate electronic properties including
+        gradients, couplings, energies. For this to work, this class
+        needs to know where the "original" files/data sit, so that they
+        can get properly passed to Turbomole. (__init__() can get these
+        file locations.)
+        """
+        self.call_and_parse_turbomole()
+
+        V = self.V(X)
+
+        self.reference, self.hamiltonian = self._compute_basis_states(self.V(X), reference=reference)
+
+        dV = self.dV(X)
+
+
+    def dV(self, X: ArrayLike) -> ArrayLike:
+        return self.gradients
+
+    def update(self, X: ArrayLike, couplings: Any = None, gradients: Any = None): 
+        out = cp.copy(self)
+        out.position = X
+        out.update_coords(X)
+        out.compute(X, couplings=couplings, gradients=gradients, reference=self.reference)
+        return out
+
+
+    def cumSH(self):
```

Comment:
This should have a more descriptive name, like `clone` or something.

Reply:
used clone